File: typescript/packages/core/src/core/databricks_volume/unlink.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { DatabricksVolumeAccessor } from '../../accessor/databricks_volume.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { record } from '../../observe/context.ts'
import { FileType, ResourceName, type PathSpec } from '../../types.ts'
import { dbxFetch } from './_client.ts'
import { ensurePathSpec } from './_helpers.ts'
import { isADirectoryError, isNotFound, notFoundError } from './errors.ts'
import { backendPath } from './path.ts'
import { stat } from './stat.ts'

export async function unlink(
  accessor: DatabricksVolumeAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<void> {
  const p = ensurePathSpec(path)
  const fileStat = await stat(accessor, p, index)
  if (fileStat.type === FileType.DIRECTORY) {
    throw isADirectoryError(p.original)
  }
  const remotePath = backendPath(accessor.config, p)
  const startMs = performance.now()
  try {
    await dbxFetch(accessor, 'DELETE', 'files', remotePath)
  } catch (exc) {
    if (isNotFound(exc)) throw notFoundError(p.original)
    throw exc
  }
  record('unlink', p.original, ResourceName.DATABRICKS_VOLUME, 0, startMs)
}

File: python/mirage/core/databricks_volume/unlink.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import time

from mirage.accessor.databricks_volume import DatabricksVolumeAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.databricks_volume._helpers import ensure_path_spec
from mirage.core.databricks_volume.errors import is_not_found
from mirage.core.databricks_volume.path import backend_path
from mirage.core.databricks_volume.stat import stat
from mirage.observe.context import record
from mirage.types import FileType, PathSpec
from mirage.utils.errors import enoent


def _delete_file_sync(
    accessor: DatabricksVolumeAccessor,
    remote_path: str,
) -> None:
    accessor.files.delete(remote_path)


async def unlink(
    accessor: DatabricksVolumeAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> None:
    path = ensure_path_spec(path)
    file_stat = await stat(accessor, path, index)
    if file_stat.type == FileType.DIRECTORY:
        raise IsADirectoryError(path.original)
    remote_path = backend_path(accessor.config, path)
    start_ms = int(time.monotonic() * 1000)
    try:
        await asyncio.to_thread(_delete_file_sync, accessor, remote_path)
    except Exception as exc:
        if is_not_found(exc):
            raise enoent(path) from exc
        raise
    record("unlink", path.original, "databricks_volume", 0, start_ms)
